PROJECT_MCPU/MCPU/DOC/GENERATED/man/man3/CanSimulator.3
.TH "CanSimulator" 3 "MCPU" \" -*- nroff -*-
.ad l
.nh
.SH NAME
CanSimulator \- This Class implements the Tcp/Ip Client connection with the SmartHub server\&.  

.SH SYNOPSIS
.br
.PP
.PP
\fR#include <CanSimulator\&.h>\fP
.SS "Public Types"

.in +1c
.ti -1c
.RI "enum class \fBcan_buf_struct\fP { \fBSTX\fP = 0, \fBLENGHT\fP, \fBCAN_IDL\fP, \fBCAN_IDH\fP, \fBDATALEN\fP, \fBD0\fP, \fBD1\fP, \fBD2\fP, \fBD3\fP, \fBD4\fP, \fBD5\fP, \fBD6\fP, \fBD7\fP, \fBETX\fP, \fBBUFLEN\fP }"
.br
.in -1c
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "delegate void \fBrxData_slot\fP (void)"
.br
.ti -1c
.RI "delegate void \fBconnection_slot\fP (bool status)"
.br
.RI "This is the delegate of the connection handler\&. "
.ti -1c
.RI "\fBCanSimulator\fP (void)"
.br
.RI "This is the class constructor\&. "
.in -1c
.SS "Static Public Member Functions"

.in +1c
.ti -1c
.RI "static bool \fBsend\fP (unsigned short \fBcanId\fP, unsigned char *data, unsigned char len)"
.br
.RI "This is function to send data in the BUS\&. "
.ti -1c
.RI "static void \fBsendMotorRotConfiguration\fP (unsigned char devId, double rot_convertion)"
.br
.ti -1c
.RI "static void \fBsendFilterConfiguration\fP (void)"
.br
.in -1c
.SS "Public Attributes"

.in +1c
.ti -1c
.RI "literal unsigned char \fBSIMUL_CONFIG_ACTIVE_DEVICES\fP = 0"
.br
.ti -1c
.RI "literal unsigned char \fBSIMUL_CONFIG_MOTOR_CONFIG_TRX\fP = 3"
.br
.ti -1c
.RI "literal unsigned char \fBSIMUL_CONFIG_MOTOR_CONFIG_ARM\fP = 4"
.br
.ti -1c
.RI "literal unsigned char \fBSIMUL_CONFIG_MOTOR_CONFIG_SLIDE\fP = 5"
.br
.ti -1c
.RI "literal unsigned char \fBSIMUL_CONFIG_MOTOR_CONFIG_BODY\fP = 6"
.br
.ti -1c
.RI "literal unsigned char \fBSIMUL_CONFIG_MOTOR_CONFIG_VERTICAL\fP = 7"
.br
.ti -1c
.RI "literal unsigned char \fBSIMUL_CONFIG_FILTERS\fP = 10"
.br
.in -1c
.SS "Static Public Attributes"

.in +1c
.ti -1c
.RI "static const System::Byte \fBSTX\fP = 0x3"
.br
.ti -1c
.RI "static const System::Byte \fBETX\fP = 0x4"
.br
.ti -1c
.RI "static \fBCanSimulator\fP ^ \fBdevice\fP = gcnew \fBCanSimulator\fP()"
.br
.ti -1c
.RI "static unsigned char \fBdatalen\fP"
.br
.ti -1c
.RI "static unsigned short \fBcanId\fP"
.br
.ti -1c
.RI "static cli::array< System::Byte > ^ \fBcanDataBuffer\fP"
.br
.RI "This is the data buffer decoded\&. "
.ti -1c
.RI "static bool \fBconnection_status\fP = false"
.br
.RI "This is the current connection status\&. "
.in -1c
.SS "Events"

.in +1c
.ti -1c
.RI "static \fBrxData_slot\fP^ \fBcanrx_canopen_sdo_event\fP"
.br
.RI "CanOpen SDO received buffer\&. "
.ti -1c
.RI "static \fBrxData_slot\fP^ \fBcanrx_canopen_bootup_event\fP"
.br
.ti -1c
.RI "static \fBrxData_slot\fP^ \fBcanrx_device_event\fP"
.br
.RI "Metaltronica Buffer\&. "
.ti -1c
.RI "static \fBconnection_slot\fP^ \fBconnection_event\fP"
.br
.RI "Connection event\&. "
.in -1c
.SS "Static Private Member Functions"

.in +1c
.ti -1c
.RI "static void \fBthreadWork\fP (void)"
.br
.RI "This is the worker thread for the connection management\&. "
.ti -1c
.RI "static void \fBhandleBuffer\fP (void)"
.br
.ti -1c
.RI "static void \fBsendConfiguration\fP (void)"
.br
.in -1c
.SS "Static Private Attributes"

.in +1c
.ti -1c
.RI "static bool \fBstarted\fP = false"
.br
.ti -1c
.RI "static System::Threading::Thread ^ \fBrunning_thread\fP"
.br
.RI "This is the worker thread handler\&. "
.ti -1c
.RI "static System::Net::Sockets::Socket ^ \fBserverSocket\fP"
.br
.RI "This is the server socket handler\&. "
.ti -1c
.RI "static System::Net::Sockets::Socket ^ \fBclientSocket\fP"
.br
.RI "This is the connected client socket handler\&. "
.ti -1c
.RI "static int \fBrx_rc\fP"
.br
.RI "This is the number of the received bytes\&. "
.ti -1c
.RI "static cli::array< System::Byte > ^ \fBrxBuffer\fP"
.br
.RI "This is the reception buffer\&. "
.in -1c
.SH "Detailed Description"
.PP 
This Class implements the Tcp/Ip Client connection with the SmartHub server\&. 


.br
 
.SH "Member Enumeration Documentation"
.PP 
.SS "enum class \fBCanSimulator::can_buf_struct\fP\fR [strong]\fP"

.PP
\fBEnumerator\fP
.in +1c
.TP
\f(BISTX \fP(0)
.TP
\f(BILENGHT \fP
.TP
\f(BICAN_IDL \fP
.TP
\f(BICAN_IDH \fP
.TP
\f(BIDATALEN \fP
.TP
\f(BID0 \fP
.TP
\f(BID1 \fP
.TP
\f(BID2 \fP
.TP
\f(BID3 \fP
.TP
\f(BID4 \fP
.TP
\f(BID5 \fP
.TP
\f(BID6 \fP
.TP
\f(BID7 \fP
.TP
\f(BIETX \fP
.TP
\f(BIBUFLEN \fP
.SH "Constructor & Destructor Documentation"
.PP 
.SS "CanSimulator::CanSimulator (void )"

.PP
This is the class constructor\&. This is the Class constructor\&.

.PP
The constructor creates all the data necessaries to create the server socket\&.

.PP
After the data initialization the constructor runs the worker thread for the Listening and data reception\&. The thread is self destroying when the program closes\&.

.PP
\fBParameters\fP
.RS 4
\fIip\fP This is a String for the IP address
.br
\fIport\fP This is an integer value for thr server port
.RE
.PP

.SH "Member Function Documentation"
.PP 
.SS "delegate void CanSimulator::connection_slot (bool status)"

.PP
This is the delegate of the connection handler\&. 
.SS "void CanSimulator::handleBuffer (void )\fR [static]\fP, \fR [private]\fP"

.SS "delegate void CanSimulator::rxData_slot (void )"

.SS "bool CanSimulator::send (unsigned short canId, unsigned char * data, unsigned char len)\fR [static]\fP"

.PP
This is function to send data in the BUS\&. 
.SS "void CanSimulator::sendConfiguration (void )\fR [static]\fP, \fR [private]\fP"

.SS "void CanSimulator::sendFilterConfiguration (void )\fR [static]\fP"

.SS "void CanSimulator::sendMotorRotConfiguration (unsigned char devId, double rot_convertion)\fR [static]\fP"

.SS "void CanSimulator::threadWork (void )\fR [static]\fP, \fR [private]\fP"

.PP
This is the worker thread for the connection management\&. 
.SH "Member Data Documentation"
.PP 
.SS "cli::array<System::Byte> ^ CanSimulator::canDataBuffer\fR [static]\fP"

.PP
This is the data buffer decoded\&. 
.SS "unsigned short CanSimulator::canId\fR [static]\fP"

.SS "System::Net::Sockets::Socket ^ CanSimulator::clientSocket\fR [static]\fP, \fR [private]\fP"

.PP
This is the connected client socket handler\&. 
.SS "bool CanSimulator::connection_status = false\fR [static]\fP"

.PP
This is the current connection status\&. 
.SS "unsigned char CanSimulator::datalen\fR [static]\fP"

.SS "\fBCanSimulator\fP ^ CanSimulator::device = gcnew \fBCanSimulator\fP()\fR [static]\fP"

.SS "const System::Byte CanSimulator::ETX = 0x4\fR [static]\fP"

.SS "System::Threading::Thread ^ CanSimulator::running_thread\fR [static]\fP, \fR [private]\fP"

.PP
This is the worker thread handler\&. 
.SS "int CanSimulator::rx_rc\fR [static]\fP, \fR [private]\fP"

.PP
This is the number of the received bytes\&. 
.SS "cli::array<System::Byte> ^ CanSimulator::rxBuffer\fR [static]\fP, \fR [private]\fP"

.PP
This is the reception buffer\&. 
.SS "System::Net::Sockets::Socket ^ CanSimulator::serverSocket\fR [static]\fP, \fR [private]\fP"

.PP
This is the server socket handler\&. 
.SS "literal unsigned char CanSimulator::SIMUL_CONFIG_ACTIVE_DEVICES = 0"

.SS "literal unsigned char CanSimulator::SIMUL_CONFIG_FILTERS = 10"

.SS "literal unsigned char CanSimulator::SIMUL_CONFIG_MOTOR_CONFIG_ARM = 4"

.SS "literal unsigned char CanSimulator::SIMUL_CONFIG_MOTOR_CONFIG_BODY = 6"

.SS "literal unsigned char CanSimulator::SIMUL_CONFIG_MOTOR_CONFIG_SLIDE = 5"

.SS "literal unsigned char CanSimulator::SIMUL_CONFIG_MOTOR_CONFIG_TRX = 3"

.SS "literal unsigned char CanSimulator::SIMUL_CONFIG_MOTOR_CONFIG_VERTICAL = 7"

.SS "bool CanSimulator::started = false\fR [static]\fP, \fR [private]\fP"

.SS "const System::Byte CanSimulator::STX = 0x3\fR [static]\fP"

.SH "Event Documentation"
.PP 
.SS "\fBrxData_slot\fP^ CanSimulator::canrx_canopen_bootup_event\fR [static]\fP"

.SS "\fBrxData_slot\fP^ CanSimulator::canrx_canopen_sdo_event\fR [static]\fP"

.PP
CanOpen SDO received buffer\&. 
.SS "\fBrxData_slot\fP^ CanSimulator::canrx_device_event\fR [static]\fP"

.PP
Metaltronica Buffer\&. 
.SS "\fBconnection_slot\fP^ CanSimulator::connection_event\fR [static]\fP"

.PP
Connection event\&. 

.SH "Author"
.PP 
Generated automatically by Doxygen for MCPU from the source code\&.


PROJECT_MCPU/MCPU/MCPU_MASTER/VSCAN/CanSimulator.cpp

#include <thread>
#include "CanSimulator.h"
#include "mutex"
#include "Log.h"
#include "ConfigurationFiles.h"

#include "PCB301.h"
#include "PCB302.h"
#include "PCB303.h"
#include "PCB304.h"
#include "PCB326.h"
#include "ArmMotor.h"
#include "TiltMotor.h"
#include "SlideMotor.h"
#include "BodyMotor.h"
#include "VerticalMotor.h"


using namespace System;
using namespace System::Net::Sockets;
using namespace System::Net;
using namespace System::Threading;
using namespace System::Diagnostics;

static std::mutex send_mutex;

/// <summary>
/// This is the Class constructor 
/// 
/// The constructor creates all the data necessaries to create the server socket.
/// 
/// After the data initialization the constructor runs the worker thread 
/// for the Listening and data reception. The thread is self destroying when the 
/// program closes.
/// 
/// 
/// </summary>
/// <param name="ip">This is a String for the IP address</param>
/// <param name="port">This is an integer value for thr server port</param>
CanSimulator::CanSimulator(void)
{
	// Gets the Handler of the Form parent class, to be used for the message exchange
	connection_status = false;
	started = false;

	canDataBuffer = gcnew cli::array<Byte>(8);

	// Sets the Tcp/Ip data for the socket
	IPAddress^ localAddress = IPAddress::Parse(SystemConfig::Configuration->getParam(SystemConfig::PARAM_SYM_MODE)[SystemConfig::SYM_MODE_IP]);
	SocketType sockType = SocketType::Stream;
	ProtocolType sockProtocol = ProtocolType::Tcp;
	serverSocket = nullptr;


	IPEndPoint^ localEndPoint = gcnew IPEndPoint(localAddress, System::Convert::ToInt16(SystemConfig::Configuration->getParam(SystemConfig::PARAM_SYM_MODE)[SystemConfig::SYM_MODE_PORT]));
	IPEndPoint^ senderAddress = gcnew IPEndPoint(localAddress, 0);

	clientSocket = nullptr;
	rxBuffer = gcnew cli::array<System::Byte>(4096);
	rx_rc = 0;

	// Create the server socket
	serverSocket = gcnew Socket(localAddress->AddressFamily, sockType, sockProtocol);
	serverSocket->Bind(localEndPoint);

	// Start the reception thread
	running_thread = gcnew Thread(gcnew ThreadStart(&CanSimulator::threadWork));
	running_thread->Name = "Loop Simulator Server";
	running_thread->IsBackground = true; // Important!!! This is necessary to allow the thread to exit when the program exits !!!
	running_thread->Start();
}

void CanSimulator::threadWork(void) {
	DWORD dwWaitResult;

	while (true) {
		serverSocket->Listen(5);
		clientSocket = serverSocket->Accept();
		LogClass::logInFile("Simulator Server: Connected!\n");
		connection_status = true;

		// Notifies the connection status
		connection_event(true);

		// Sends the system configuration to the simulator application
		sendConfiguration();

		while (true) {
			try {
				rx_rc = clientSocket->Receive(rxBuffer);
				if (rx_rc == 0) break; // Shutdown
				handleBuffer();
				
			}
			catch (...) {
				break;
			}

		}
		connection_status = false;

		// The Client closed the connection
		LogClass::logInFile("Simulator Server: Shutdown!\n");
		clientSocket->Shutdown(SocketShutdown::Send);
		clientSocket->Close();

		// Notifies the disconnection status
		connection_event(false);
	}

}
void CanSimulator::sendConfiguration() {
	cli::array<Byte>^ buffer = gcnew cli::array<Byte>((int)can_buf_struct::BUFLEN);

	buffer[(Byte)can_buf_struct::STX] = CanSimulator::STX;
	buffer[(Byte)can_buf_struct::LENGHT] = (Byte)can_buf_struct::BUFLEN;
	buffer[(Byte)can_buf_struct::CAN_IDL] = SIMUL_CONFIG_ACTIVE_DEVICES;
	buffer[(Byte)can_buf_struct::CAN_IDH] = 0xFF;
	buffer[(Byte)can_buf_struct::ETX] = CanSimulator::ETX;
	buffer[(Byte)can_buf_struct::DATALEN] = 8;
	
	// Board Simulation bit
	buffer[(Byte)can_buf_struct::D0] = 0;
	if (Gantry::isPcb301Demo()) buffer[(Byte)can_buf_struct::D0] |= 1;
	if (Gantry::isPcb302Demo()) buffer[(Byte)can_buf_struct::D0] |= 2;
	if (Gantry::isPcb303Demo()) buffer[(Byte)can_buf_struct::D0] |= 4;
	if (Gantry::isPcb304Demo()) buffer[(Byte)can_buf_struct::D0] |= 8;
	if (Gantry::isPcb325Demo()) buffer[(Byte)can_buf_struct::D0] |= 0x10;
	if (Gantry::isPcb326Demo()) buffer[(Byte)can_buf_struct::D0] |= 0x20;

	// Motor Simulation Bit
	buffer[(Byte)can_buf_struct::D1] = 0;
	if (Gantry::isMotorVerticalDemo()) buffer[(Byte)can_buf_struct::D1] |= 1;
	if (Gantry::isMotorBodyDemo()) buffer[(Byte)can_buf_struct::D1] |= 2;
	if (Gantry::isMotorArmDemo()) buffer[(Byte)can_buf_struct::D1] |= 4;
	if (Gantry::isMotorTiltDemo()) buffer[(Byte)can_buf_struct::D1] |= 8;
	if (Gantry::isMotorSlideDemo()) buffer[(Byte)can_buf_struct::D1] |= 0x10;
	
	try {
		device->clientSocket->Send(buffer);
	}
	catch (...) {
		LogClass::logInFile("Simulator Server: failed sending the configuration word !\n");
	}

	return;
}

void CanSimulator::sendMotorRotConfiguration(unsigned char devId, double rot_convertion) {
	
	cli::array<System::Byte>^ buffer = gcnew cli::array<System::Byte>((int)can_buf_struct::BUFLEN);
	char* byteArray = reinterpret_cast<char*>(&rot_convertion);
	for (int i = 0; i < 8; i++) buffer[(Byte)can_buf_struct::D0 + i] = byteArray[i];

	buffer[(System::Byte)can_buf_struct::STX] = CanSimulator::STX;
	buffer[(System::Byte)can_buf_struct::LENGHT] = (Byte)can_buf_struct::BUFLEN;
	buffer[(System::Byte)can_buf_struct::CAN_IDL] = devId;
	buffer[(System::Byte)can_buf_struct::CAN_IDH] = 0xFF;
	buffer[(System::Byte)can_buf_struct::ETX] = CanSimulator::ETX;
	buffer[(Byte)can_buf_struct::DATALEN] = 8;

	try {
		device->clientSocket->Send(buffer);
	}
	catch (...) {
		LogClass::logInFile("Simulator Server: failed sending the motor configuration word !\n");
	}

	return;
}


void CanSimulator::sendFilterConfiguration(void) {

	cli::array<System::Byte>^ buffer = gcnew cli::array<System::Byte>((int)can_buf_struct::BUFLEN);
	
	
	// Assignes the Slots 
	buffer[(Byte)can_buf_struct::D0] = PCB303::getFilterSlot(PCB303::filter_index::FILTER_RH);
	buffer[(Byte)can_buf_struct::D1] = PCB303::getFilterSlot(PCB303::filter_index::FILTER_AG);
	buffer[(Byte)can_buf_struct::D2] = PCB303::getFilterSlot(PCB303::filter_index::FILTER_AL);
	buffer[(Byte)can_buf_struct::D3] = PCB303::getFilterSlot(PCB303::filter_index::FILTER_CU);
	buffer[(Byte)can_buf_struct::D4] = PCB303::getFilterSlot(PCB303::filter_index::FILTER_MO);
	buffer[(Byte)can_buf_struct::D5] = PCB303::getFilterSlot(PCB303::filter_index::FILTER_LD);

	buffer[(System::Byte)can_buf_struct::STX] = CanSimulator::STX;
	buffer[(System::Byte)can_buf_struct::LENGHT] = (Byte)can_buf_struct::BUFLEN;
	buffer[(System::Byte)can_buf_struct::CAN_IDL] = SIMUL_CONFIG_FILTERS;
	buffer[(System::Byte)can_buf_struct::CAN_IDH] = 0xFF;
	buffer[(System::Byte)can_buf_struct::ETX] = CanSimulator::ETX;
	buffer[(Byte)can_buf_struct::DATALEN] = 8;

	try {
		device->clientSocket->Send(buffer);
	}
	catch (...) {
		LogClass::logInFile("Simulator Server: failed sending the motor configuration word !\n");
	}

	return;
}


void CanSimulator::handleBuffer(void) {

	
	// Scroll for possible concatenated frames
	for (int j = 0; j < rx_rc; j++) {

		if (rx_rc - j < 4) continue;

		// Start byte
		if (rxBuffer[j + (int)can_buf_struct::STX] != (int)CanSimulator::STX)  continue;

		// Lenght of frame
		int len = rxBuffer[j + (int)can_buf_struct::LENGHT];
		if (len > rx_rc - j) continue;

		// End of frame
		if (rxBuffer[j + (int)can_buf_struct::ETX] != (int)CanSimulator::ETX) continue;

		// can Data Len
		datalen = (unsigned char)rxBuffer[j + (int)can_buf_struct::DATALEN];
		if (datalen > 8) continue;

		// Destination address
		canId = (unsigned short)rxBuffer[j + (int)can_buf_struct::CAN_IDL] + 256 * (unsigned short)rxBuffer[j + (int)can_buf_struct::CAN_IDH];

		// Destination buffer
		for (int i = 0; i < datalen; i++) canDataBuffer[i] = rxBuffer[j + (int)can_buf_struct::D0 + i];

		if ((canId >= 0x100) && (canId <= 0x17F)) {
			canrx_device_event();
			continue;
		}

		if ((canId >= 0x580) && (canId <= 0x5FF)) {

			canrx_canopen_sdo_event();
			continue;
		}

		if ((canId >= 0x700) && (canId <= 0x707)) {
			canrx_canopen_bootup_event();
			continue;
		}

	}
}


bool CanSimulator::send(unsigned short canId, unsigned char* data, unsigned char len) {
	
	if (!connection_status) return false;

	if (len > 8) len = 8;

	cli::array<Byte>^ buffer = gcnew cli::array<Byte>((int) can_buf_struct::BUFLEN);
	buffer[(Byte)can_buf_struct::STX] = CanSimulator::STX;
	buffer[(Byte)can_buf_struct::LENGHT] = (Byte)can_buf_struct::BUFLEN;
	buffer[(Byte)can_buf_struct::CAN_IDL] = (Byte) (canId & 0xff);
	buffer[(Byte)can_buf_struct::CAN_IDH] = (Byte)((canId>>8) & 0xff);
	buffer[(Byte)can_buf_struct::DATALEN] = len;
	for (unsigned char i = 0; i < len; i++) buffer[(Byte)can_buf_struct::D0 + i] = data[i];	
	buffer[(Byte)can_buf_struct::ETX] = CanSimulator::ETX;

	try{
		device->clientSocket->Send(buffer);
	}
	catch (...) {
		return false;
	}
	
	return true;
	
}

